src/demo/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtensionDemo.java

```diff
@@ -279,4 +282,30 @@ void testWithCustomDisplayName(@Values(strings = { "0", "1" }) String a,
 	}
 	// end::cartesian_testWithCustomDisplayName[]
 
+	// tag::cartesian_argument_sets_with_nested_classes[]
+	class MyCartesianTestClassWithNestedClasses {
+
+		@Nested
+		// the next annotation is required to allow a non-static factory method
+		@TestInstance(Lifecycle.PER_CLASS)
+		class MyNestedCartesianTestClass {
+
+			@CartesianTest
+			@CartesianTest.MethodFactory("provideArguments")
+			void testNeedingArguments(String string, int i) {
+				// passing test code
+			}
+
+			// this provider method cannot be static because it is in a non-static nested class
```

Comment:
Since Java 16, this is not longer correct. To make records (which are always static) easier to use in non-static nested classes, this limitation was lifted and non-static nested classes can now have static members. See [this compiling example](https://github.com/nipafx/demo-java-x/blob/main/src/main/java/dev/nipafx/demo/java16/lang/staticinner/OuterClass.java).

I think it's enough to point out that the method doesn't have to be static, but if you want, you can also add that it couldn't even be until Java 16.

Reply:
I wasn't aware of this. I've used your original suggestion.